### PR TITLE
Fix a bug caused by partial restore from shared-cache

### PR DESCRIPTION
**Pull request by aalekseyev**
This PR fixes (and adds a test demonistrating) a bug where a restoration from shared cache can partially succeed and cause the follow-up action observe some of targets as already existing, being restored from shared-cache. This is a problem because the targets restored from cache are read-only, so most actions fail when trying to overwrite them.

### CODE REVIEW

File: src/dune_cache/local.ml [hunk 1]
```diff
@@ -192,46 +192,67 @@ module Artifacts = struct
...
+                | () -> Ok (path_in_build_dir, file_digest))
+              | Copy -> (
+                match
+                  Io.copy_file ~src:path_in_cache
+                    ~dst:(Path.build path_in_build_dir)
+                    ()
+                with
+                | exception Sys_error _ -> Error Not_found
+                | () -> Ok (path_in_build_dir, file_digest)))
+            ~destroy:(fun (path_in_build_dir, _digest) ->
```
Comment (snowleopard) [inline src/dune_cache/local.ml:245]: Maybe worth moving this argument on top of the longer `create`.

**Review by snowleopard — APPROVED** [review]
Thanks for the fix, looks good!

Removing the files twice (once in `execute_rule_impl` and then again in `restore_artifacts`) doesn't seem great. I lean towards dropping the second removal, and adding a comment to `restore_artifacts` that the caller is responsible for the clean up of stale files before calling it.

File: src/dune_cache/local.ml (hunk 1 above)
Comment (aalekseyev) [reply src/dune_cache/local.ml]: Done.

**Comment by aalekseyev** [comment]
I pushed a commit removing that duplicate unlink call. Seems alright.

**Comment by snowleopard** [comment]
> I pushed a commit removing that duplicate unlink call. Seems alright.

Great, thanks!

**Pull request merged**